network.go

```diff
@@ -835,32 +835,73 @@ func (n *network) updateSvcRecord(ep *endpoint, localEps []*endpoint, isAdd bool
 	sr, ok := c.svcDb[n.ID()]
 	if !ok {
 		c.svcDb[n.ID()] = svcInfo{
-			svcMap: make(map[string]net.IP),
+			svcMap: make(map[string][]net.IP),
 			ipMap:  make(map[string]string),
 		}
 		sr = c.svcDb[n.ID()]
 	}
```

Comment:
With this change if you also move the reverseIP mapping/unmapping into `addSvcRecords` with a flag to indicate if you want to do reverseIP mapping/unmapping you can completely get rid of this code snippet of obtaining sr here since you are going to do that anyways in `addSvcRecords` and do it all at once. Also since you are properly acquiring controllerLock in `addSvcRecords` it fixes one existing race here.